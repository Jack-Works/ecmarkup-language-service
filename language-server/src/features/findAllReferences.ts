import {
    type CancellationToken,
    type Connection,
    type Location,
    Range,
    type ReferenceParams,
    type ResultProgressReporter,
    type ServerCapabilities,
    type TextDocuments,
    type WorkDoneProgressReporter,
} from 'vscode-languageserver'
import type { TextDocument } from '../lib.js'
import { word_at_cursor } from '../utils/text.js'
import type { Program } from '../workspace/program.js'

export function referenceProvider(
    connection: Connection,
    program: Program,
    documents: TextDocuments<TextDocument>,
): NonNullable<ServerCapabilities['referencesProvider']> {
    const provider = new ReferenceProvider()
    connection.onReferences(provider.handler(documents, program))
    return {}
}

export class ReferenceProvider {
    findReferences(
        document: TextDocument,
        program: Program,
        params: ReferenceParams,
        _token?: CancellationToken | undefined,
        _workDoneProgress?: WorkDoneProgressReporter | undefined,
        _resultProgress?: ResultProgressReporter<Location[]> | undefined,
    ) {
        const sourceFile = program.getSourceFile(document)
        const offset = document.offsetAt(params.position)
        const { word, isGrammarLeading, isVariable } = word_at_cursor(document.getText(), offset)
        const { includeDeclaration } = params.context

        const node = sourceFile.findNodeAt(offset)
        if (node.tag === 'emu-grammar' || isGrammarLeading) {
            const result = sourceFile.grammars.filter(
                (entry) => entry.name === word && (includeDeclaration || entry.type === 'reference'),
            )
            return result.map(
                (ref): Location => ({ uri: document.uri, range: sourceFile.getRelativeRange(ref.node, ref.range) }),
            )
        } else if (node.tag === 'h1' || node.tag === 'emu-alg') {
            if (isVariable) {
                const searchingNode = sourceFile.getAlgHeader(offset)?.parent ?? node
                const text = sourceFile.getNodeText(searchingNode)
                const location: Location[] = []
                let lastIndex = 0
                while (true) {
                    lastIndex = text.indexOf(`_${word}_`, lastIndex + 1)
                    if (lastIndex === -1) break
                    location.push({
                        uri: document.uri,
                        range: sourceFile.getRelativeRange(searchingNode, {
                            position: lastIndex + 1,
                            length: word.length,
                        }),
                    })
                }
                return location
            } else {
                const operation = sourceFile.localDefinedAbstractOperations.find((entry) => entry.name === word)
                if (operation) {
                    const text = document.getText()
                    const location: Location[] = []
                    const re =
                        'escape' in RegExp
                            ? new RegExp(
                                  `\\b${
                                      // biome-ignore lint/suspicious/noExplicitAny: remove after Node 24
                                      (RegExp.escape as any)(word)
                                  }\\b`,
                                  'gu',
                              )
                            : // bless us
                              new RegExp(`\\b${word}\\b`, 'g')
                    for (const match of text.matchAll(re)) {
                        location.push({
                            uri: document.uri,
                            range: Range.create(
                                sourceFile.text.positionAt(match.index),
                                sourceFile.text.positionAt(match.index + word.length),
                            ),
                        })
                    }
                    return location
                }
            }
        }

        return undefined
    }

    handler(documents: TextDocuments<TextDocument>, program: Program) {
        return (
            params: ReferenceParams,
            token: CancellationToken,
            workDoneProgress: WorkDoneProgressReporter,
            resultProgress: ResultProgressReporter<Location[]> | undefined,
        ) => {
            const document = documents.get(params.textDocument.uri)
            if (!document) return undefined
            return this.findReferences(document, program, params, token, workDoneProgress, resultProgress)
        }
    }
}
