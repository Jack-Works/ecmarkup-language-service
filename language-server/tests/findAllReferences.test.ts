import { expect, it } from 'vitest'
import type { Location } from 'vscode-languageserver'
import { ReferenceProvider } from '../src/features/findAllReferences.js'
import type { TextDocument } from '../src/lib.js'
import { betterSnapshot, File } from './File.js'

it('find all references', async () => {
    const goto = new ReferenceProvider()
    const { document, markers, textDocument, program } = File.of`
        <emu-grammar type="definition">
            MoreOne${File.mark('grammar')}Night ::= "more" "one" "night"
        </emu-grammar>
        <emu-prodref name="MoreOneNight"></emu-prodref>
        <p>A |MoreOneNight| production.</p>
        <emu-grammar>
            Other
                MoreOneNight
        </emu-grammar>

        <emu-clause id="sec-SaveTheWorld" type="abstract operation">
            <h1>
                ${File.mark('AO')}SaveTheWorld (
                    _input_${File.mark('parameter')}: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
                1. Let _obj_${File.mark('local variable')} be ! OrdinaryObjectCreate(_input_).
                1. Perform ? FinishSavingTheWorld(_obj_).
            </emu-alg>
        </emu-clause>
        <emu-alg>
            1. Perform ? SaveTheWorld(  ).
        </emu-alg>
    `

    for (const { position, desc } of markers) {
        const result = goto.findReferences(document, program, {
            textDocument,
            position,
            context: { includeDeclaration: true },
        })
        expect(printRefs(document, result)).toMatchSnapshot(desc)
    }
})

it('find all references no false positive (containing)', async () => {
    const goto = new ReferenceProvider()
    const { document, markers, textDocument, program } = File.of`
        <emu-clause id="sec-Save" type="abstract operation">
            <h1>
                ${File.mark()}Save (
                    _input_: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
            </emu-alg>
        </emu-clause>
        <emu-clause id="sec-SaveTheWorld" type="abstract operation">
            <h1>
                SaveTheWorld (
                    _input_: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
            </emu-alg>
        </emu-clause>
        <emu-alg>
            1. Perform ? SaveTheWorld(  ).
        </emu-alg>
    `

    for (const { position, desc } of markers) {
        const result = goto.findReferences(document, program, {
            textDocument,
            position,
            context: { includeDeclaration: true },
        })
        expect(printRefs(document, result)).toMatchSnapshot(desc)
    }
})

it('find all references no false positive (cross AO variable)', async () => {
    const goto = new ReferenceProvider()
    const { document, markers, textDocument, program } = File.of`
        <emu-clause id="sec-Save" type="abstract operation">
            <h1>
                Save (
                    _input_${File.mark()}: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
                1. Set _input_ to *undefined*.
            </emu-alg>
        </emu-clause>
        <emu-clause id="sec-SaveTheWorld" type="abstract operation">
            <h1>
                SaveTheWorld (
                    _input_: an ECMAScript language value,
                ): either a normal completion containing an Object or a throw completion
            </h1>
            <dl class="header">
            </dl>
            <emu-alg>
                1. Set _input_ to *undefined*.
            </emu-alg>
        </emu-clause>
        <emu-alg>
            1. Perform ? SaveTheWorld(  ).
        </emu-alg>
    `

    for (const { position, desc } of markers) {
        const result = goto.findReferences(document, program, {
            textDocument,
            position,
            context: { includeDeclaration: true },
        })
        expect(printRefs(document, result)).toMatchSnapshot(desc)
    }
})

function printRefs(document: TextDocument, tokens: Location[] | undefined) {
    if (tokens === undefined) return 'No reference found'
    return betterSnapshot(
        document,
        tokens.map(({ range }) => ({ annotate: '', range })),
    )
}
