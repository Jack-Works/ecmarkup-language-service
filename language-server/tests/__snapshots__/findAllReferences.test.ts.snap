// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`find all references > AO 1`] = `
"<emu-grammar type="definition">
    MoreOneNight ::= "more" "one" "night"
</emu-grammar>
<emu-prodref name="MoreOneNight"></emu-prodref>
<p>A |MoreOneNight| production.</p>
<emu-grammar>
    Other
        MoreOneNight
</emu-grammar>

<emu-clause id="sec-SaveTheWorld" type="abstract operation">
                    ~~~~~~~~~~~~
    <h1>
        SaveTheWorld (
        ~~~~~~~~~~~~
            _input_: an ECMAScript language value,
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
        1. Let _obj_ be ! OrdinaryObjectCreate(_input_).
        1. Perform ? FinishSavingTheWorld(_obj_).
    </emu-alg>
</emu-clause>
<emu-alg>
    1. Perform ? SaveTheWorld(  ).
                 ~~~~~~~~~~~~
</emu-alg>"
`;

exports[`find all references > grammar 1`] = `
"<emu-grammar type="definition">
    MoreOneNight ::= "more" "one" "night"
    ~~~~~~~~~~~~
</emu-grammar>
<emu-prodref name="MoreOneNight"></emu-prodref>
                   ~~~~~~~~~~~~
<p>A |MoreOneNight| production.</p>
      ~~~~~~~~~~~~
<emu-grammar>
    Other
        MoreOneNight
        ~~~~~~~~~~~~
</emu-grammar>

<emu-clause id="sec-SaveTheWorld" type="abstract operation">
    <h1>
        SaveTheWorld (
            _input_: an ECMAScript language value,
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
        1. Let _obj_ be ! OrdinaryObjectCreate(_input_).
        1. Perform ? FinishSavingTheWorld(_obj_).
    </emu-alg>
</emu-clause>
<emu-alg>
    1. Perform ? SaveTheWorld(  ).
</emu-alg>"
`;

exports[`find all references > local variable 1`] = `
"<emu-grammar type="definition">
    MoreOneNight ::= "more" "one" "night"
</emu-grammar>
<emu-prodref name="MoreOneNight"></emu-prodref>
<p>A |MoreOneNight| production.</p>
<emu-grammar>
    Other
        MoreOneNight
</emu-grammar>

<emu-clause id="sec-SaveTheWorld" type="abstract operation">
    <h1>
        SaveTheWorld (
            _input_: an ECMAScript language value,
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
        1. Let _obj_ be ! OrdinaryObjectCreate(_input_).
                ~~~
        1. Perform ? FinishSavingTheWorld(_obj_).
                                           ~~~
    </emu-alg>
</emu-clause>
<emu-alg>
    1. Perform ? SaveTheWorld(  ).
</emu-alg>"
`;

exports[`find all references > parameter 1`] = `
"<emu-grammar type="definition">
    MoreOneNight ::= "more" "one" "night"
</emu-grammar>
<emu-prodref name="MoreOneNight"></emu-prodref>
<p>A |MoreOneNight| production.</p>
<emu-grammar>
    Other
        MoreOneNight
</emu-grammar>

<emu-clause id="sec-SaveTheWorld" type="abstract operation">
    <h1>
        SaveTheWorld (
            _input_: an ECMAScript language value,
             ~~~~~
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
        1. Let _obj_ be ! OrdinaryObjectCreate(_input_).
                                                ~~~~~
        1. Perform ? FinishSavingTheWorld(_obj_).
    </emu-alg>
</emu-clause>
<emu-alg>
    1. Perform ? SaveTheWorld(  ).
</emu-alg>"
`;

exports[`find all references no false positive (containing) 1`] = `
"<emu-clause id="sec-Save" type="abstract operation">
                    ~~~~
    <h1>
        Save (
        ~~~~
            _input_: an ECMAScript language value,
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
    </emu-alg>
</emu-clause>
<emu-clause id="sec-SaveTheWorld" type="abstract operation">
    <h1>
        SaveTheWorld (
            _input_: an ECMAScript language value,
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
    </emu-alg>
</emu-clause>
<emu-alg>
    1. Perform ? SaveTheWorld(  ).
</emu-alg>"
`;

exports[`find all references no false positive (cross AO variable) 1`] = `
"<emu-clause id="sec-Save" type="abstract operation">
    <h1>
        Save (
            _input_: an ECMAScript language value,
             ~~~~~
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
        1. Set _input_ to *undefined*.
                ~~~~~
    </emu-alg>
</emu-clause>
<emu-clause id="sec-SaveTheWorld" type="abstract operation">
    <h1>
        SaveTheWorld (
            _input_: an ECMAScript language value,
        ): either a normal completion containing an Object or a throw completion
    </h1>
    <dl class="header">
    </dl>
    <emu-alg>
        1. Set _input_ to *undefined*.
    </emu-alg>
</emu-clause>
<emu-alg>
    1. Perform ? SaveTheWorld(  ).
</emu-alg>"
`;
